### PR TITLE
Allow changing of CPUs, Memory, and Disk Size

### PR DESCRIPTION
**Pull request by ashley-cui**
Allow podman machine set to change CPUs, Memory and Disk size of a QEMU machine after its been created.
Disk size can only be increased.

If one setting fails to be changed, the other settings will still be applied.

Signed-off-by: Ashley Cui <acui@redhat.com>

Fixes: https://github.com/containers/podman/issues/13633

### CODE REVIEW

File: cmd/podman/machine/set.go [hunk 1]
```diff
@@ -23,9 +26,17 @@ var (
 )
 
 var (
-	setOpts = machine.SetOptions{}
+	setFlags = SetFlags{}
+	setOpts  = machine.SetOptions{}
 )
 
+type SetFlags struct {
+	Rootful  bool
```
Comment (rhatdan) [inline cmd/podman/machine/set.go:34]: nit: Spell this with Rootfull

File: cmd/podman/machine/set.go [hunk 2]
```diff
@@ -53,5 +90,26 @@ func setMachine(cmd *cobra.Command, args []string) error {
...
+		setOpts.Memory = &setFlags.Memory
+		fmt.Println("memory")
+	}
+	if cmd.Flags().Changed("disk-size") {
+		setOpts.DiskSize = &setFlags.DiskSize
+	}
+
+	setErrs, err := vm.Set(vmName, setOpts)
+
+	if len(setErrs) > 0 {
```
Comment (rhatdan) [inline cmd/podman/machine/set.go:109]: No need for this check.

Comment (rhatdan) [reply cmd/podman/machine/set.go:109]: I would check for err != nil and return it, and then go through the list of setErrs?  Or do they get set and return err != nil?

File: pkg/machine/qemu/machine.go [hunk 3]
```diff
@@ -430,45 +409,77 @@ func (v *MachineVM) Init(opts machine.InitOptions) (bool, error) {
...
+		}
+		if !found {
+			v.CmdLine = append(v.CmdLine, []string{"-m", strconv.Itoa(int(v.Memory))}...)
+		}
+	}
+
+	if opts.DiskSize != nil && v.DiskSize != *opts.DiskSize {
+		err := v.resizeDisk(v.DiskSize)
+		if err != nil {
+			setErrors = append(setErrors, errors.Wrapf(err, "error resizing disk"))
```
Comment (rhatdan) [inline pkg/machine/qemu/machine.go:476]: Remove "error "; Podman prepends "Error:"

File: pkg/machine/qemu/machine.go [hunk 4]
```diff
@@ -430,45 +409,77 @@ func (v *MachineVM) Init(opts machine.InitOptions) (bool, error) {
...
+		return setErrors, errors.Errorf("cannot change settings while the vm is running, run 'podman machine stop%s' first", suffix)
 	}
 
-	changeCon, err := machine.AnyConnectionDefault(v.Name, v.Name+"-root")
-	if err != nil {
-		return err
+	if opts.Rootful != nil && v.Rootful == *opts.Rootful {
+		err := v.setRootful(*opts.Rootful)
+		if err != nil {
+			setErrors = append(setErrors, errors.Wrapf(err, "error setting rootful option"))
```
Comment (rhatdan) [inline pkg/machine/qemu/machine.go:437]: Remove "error "; Podman prepends "Error:"

File: cmd/podman/machine/set.go (hunk 2 above)
Comment (ashley-cui) [reply cmd/podman/machine/set.go:109]: setErrors are errors while setting the settings. There can be more than one error, or none at all, ie, if changing --cpus and --memory both error, then both errors will be in the setErrs list. 

The other err is if writing the machine config causes an error. This a little worse because it might mean settings are set but not updated in the machine config.

File: pkg/machine/qemu/machine.go [hunk 5]
```diff
@@ -430,45 +409,75 @@ func (v *MachineVM) Init(opts machine.InitOptions) (bool, error) {
...
+				break
+			}
+		}
+		if !found {
+			v.CmdLine = append(v.CmdLine, []string{"-m", strconv.Itoa(int(v.Memory))}...)
+		}
+	}
+
+	if opts.DiskSize != nil && v.DiskSize != *opts.DiskSize {
+		if err := v.resizeDisk(v.DiskSize); err != nil {
```
Comment (Luap99) [inline pkg/machine/qemu/machine.go:473]: Does this actually resize the filesystem? I suspect that you have to resize the filesystem in the VM. Resizing only the disk is pointless.

File: pkg/machine/qemu/machine.go [hunk 6]
```diff
@@ -430,45 +409,75 @@ func (v *MachineVM) Init(opts machine.InitOptions) (bool, error) {
...
+		for i, val := range v.CmdLine {
+			if val == "-m" {
+				found = true
+				v.CmdLine[i+1] = strconv.Itoa(int(v.Memory))
+				break
+			}
+		}
+		if !found {
+			v.CmdLine = append(v.CmdLine, []string{"-m", strconv.Itoa(int(v.Memory))}...)
+		}
```
Comment (Luap99) [inline pkg/machine/qemu/machine.go:469]: Can we turn this into a function since you use the same logic at least twice.

File: pkg/machine/qemu/machine.go [hunk 7]
```diff
@@ -1411,3 +1420,53 @@ func (v *MachineVM) getImageFile() string {
...
+
+func (v *MachineVM) resizeDisk(diskSize uint64) error {
+	originalDiskSize, err := getDiskSize(v.getImageFile())
+	if err != nil {
+		return err
+	}
+	// Resize the disk image to input disk size
+	// only if the virtualdisk size is less than
+	// the given disk size
+	if diskSize<<(10*3) <= originalDiskSize {
```
Comment (Luap99) [inline pkg/machine/qemu/machine.go:1432]: Can you add a comment what this <<30 should do. I assume it converts bytes to mb but a comment should help.

File: cmd/podman/machine/set.go [hunk 8]
```diff
@@ -53,5 +89,22 @@ func setMachine(cmd *cobra.Command, args []string) error {
...
+		setOpts.Memory = &setFlags.Memory
+	}
+	if cmd.Flags().Changed("disk-size") {
+		setOpts.DiskSize = &setFlags.DiskSize
+	}
+
+	setErrs, err := vm.Set(vmName, setOpts)
+	for _, err := range setErrs {
+		fmt.Fprintf(os.Stderr, "%v\n", err)
+	}
```
Comment (Luap99) [inline cmd/podman/machine/set.go:108]: I think we already have a package for this
https://github.com/containers/podman/blob/main/cmd/podman/utils/error.go

File: cmd/podman/machine/set.go [hunk 9]
```diff
@@ -34,7 +45,33 @@ func init() {
 	flags := setCmd.Flags()
 
 	rootfulFlagName := "rootful"
-	flags.BoolVar(&setOpts.Rootful, rootfulFlagName, false, "Whether this machine should prefer rootful container execution")
+	flags.BoolVar(&setFlags.Rootful, rootfulFlagName, false, "Whether this machine should prefer rootful container execution")
```
[the remark this answers is not in the document]
Comment (TomSweeneyRedHat) [reply cmd/podman/machine/set.go:48]: I too think `rootfull` is more correct, but I  that ship has all but sailed.  We've already established `rootful` in a number of places, including RHEL docs.

File: cmd/podman/machine/set.go [hunk 10]
```diff
@@ -23,9 +26,17 @@ var (
...
-	setOpts = machine.SetOptions{}
+	setFlags = SetFlags{}
+	setOpts  = machine.SetOptions{}
 )
 
+type SetFlags struct {
+	Rootful  bool
+	CPUs     uint64
+	DiskSize uint64
+	Memory   uint64
```
Comment (TomSweeneyRedHat) [inline cmd/podman/machine/set.go:37]: My OCD would be much happier if these were in alpha order.

File: cmd/podman/machine/set.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (rhatdan) [reply cmd/podman/machine/set.go:48]: The man pages have both, I will open a PR to fix this and alias the --rootful and --rootfull options in podman machine.
Lets still to rootfull since it line up well with rootless.

File: docs/source/markdown/podman-machine-set.1.md [hunk 11]
```diff
@@ -8,17 +8,29 @@ podman\-machine\-set - Sets a virtual machine setting
...
 Print usage statement.
 
+#### **--memory**, **-m**=*number*
+
+Memory (in MB).
+Only supported for QEMU machines.
+
 #### **--rootful**=*true|false*
 
 Whether this machine should prefer rootful (`true`) or rootless (`false`)
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-machine-set.1.md:36]: I know not part of this change, but it looks like there's an extra space in front of the first word in this line.  Also, the default value is not listed.  Is it true or false, or will the machine not start without this option?

File: pkg/machine/config.go [hunk 12]
```diff
@@ -95,7 +95,10 @@ type ListResponse struct {
 }
 
 type SetOptions struct {
-	Rootful bool
+	Rootful  *bool
+	CPUs     *uint64
+	DiskSize *uint64
+	Memory   *uint64
```
Comment (TomSweeneyRedHat) [inline pkg/machine/config.go:101]: OCD alpha time again

File: cmd/podman/machine/set.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (rhatdan) [reply cmd/podman/machine/set.go:48]: https://github.com/containers/podman/pull/13964

File: docs/source/markdown/podman-machine-set.1.md [hunk 13]
```diff
@@ -8,17 +8,29 @@ podman\-machine\-set - Sets a virtual machine setting
...
+Can only be increased. Only supported for QEMU machines.
+
 #### **--help**
 
 Print usage statement.
 
+#### **--memory**, **-m**=*number*
+
+Memory (in MB).
+Only supported for QEMU machines.
```
[the remark this answers is not in the document]
Comment (rhatdan) [reply docs/source/markdown/podman-machine-set.1.md:32]: Do these man pages point out that these will not work on Windows platforms?

File: docs/source/markdown/podman-machine-set.1.md (hunk 11 above)
Comment (ashley-cui) [reply docs/source/markdown/podman-machine-set.1.md:36]: Will remove the extra space, thanks!

The default value will be set in `podman machine init`. This is a the `set` command, which changes machine settings after the machine is created. There is no default value; if the flag is not explicitly used, the memory of the machine will not change.

File: cmd/podman/machine/set.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (ashley-cui) [reply cmd/podman/machine/set.go:48]: https://github.com/containers/podman/pull/13964#issuecomment-1106546414

[the remark this answers is not in the document]
Comment (ashley-cui) [reply cmd/podman/machine/set.go:48]: I'd also be in favor of dropping the suffix entirely, and changing the setting to --root, or --connection=root|rootless

[the remark this answers is not in the document]
Comment (rhatdan) [reply cmd/podman/machine/set.go:48]: I would be fine with switching it to --root, but `podman --root` has this option.

Switching the option to a param is not a great solution since the option already exists.

File: docs/source/markdown/podman-machine-set.1.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (ashley-cui) [reply docs/source/markdown/podman-machine-set.1.md:32]: Yeah, windows assume the type of the machine is WSL, this mentions that only qemu machines are supported.

**Comment by baude** [comment]
I didn't see any error like `error: new disk size must be larger than current disk size %q` or something similar.  Did I miss it?

File: docs/source/markdown/podman-machine-set.1.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (ashley-cui) [reply docs/source/markdown/podman-machine-set.1.md:32]: I think the qemu manpage listing should be unnecessary we're not necessarily talking about qemu, the binary, more so `qemu` as a type of podman-machine.

qemu(1) also doesn't exist

File: pkg/machine/qemu/machine.go (hunk 5 above)
Comment (ashley-cui) [reply pkg/machine/qemu/machine.go:473]: [core@localhost ~]$ sudo fdisk -l | grep Disk
Disk /dev/vda: 150 GiB, 161061273600 bytes, 314572800 sectors
Disklabel type: gpt
Disk identifier: 23EB28A4-C273-11EC-B7E5-01BDD81662FF
[core@localhost ~]$ df -h
Filesystem      Size  Used Avail Use% Mounted on
devtmpfs        4.0M     0  4.0M   0% /dev
tmpfs           979M  100K  979M   1% /dev/shm
tmpfs           392M  5.7M  386M   2% /run
/dev/vda4       150G  2.5G  148G   2% /sysroot
overlay         150G  2.5G  148G   2% /usr
tmpfs           979M     0  979M   0% /tmp
/dev/vda3       362M  110M  230M  33% /boot
tmpfs           196M  8.0K  196M   1% /run/user/501
vol0            229G  113G  117G  50% /Users/acui


Not an expert here but it looks like it resized the fs?

**Comment by ashley-cui** [comment]
@baude bug should be fixed, thanks!

File: cmd/podman/machine/set.go (hunk 8 above)
Comment (ashley-cui) [reply cmd/podman/machine/set.go:108]: Looks like that type is in `cmd/podman`, and I don't want to import `cmd/podman` into `pkg/machine`

**Comment by ashley-cui** [comment]
Holding until https://github.com/containers/podman/pull/13964#issuecomment-1106689431 is resolved.

**Comment by rhatdan** [comment]
LGTM

File: pkg/machine/qemu/machine.go [hunk 14]
```diff
@@ -1411,3 +1405,64 @@ func (v *MachineVM) getImageFile() string {
...
+		}
+		err := machine.ChangeDefault(newDefault)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (v *MachineVM) editCmdLine(flag string, value string) {
```
Comment (baude) [inline pkg/machine/qemu/machine.go:1456]: ripe for a unittest?

File: pkg/machine/qemu/machine.go [hunk 15]
```diff
@@ -1411,3 +1405,64 @@ func (v *MachineVM) getImageFile() string {
...
+	resize.Stdout = os.Stdout
+	resize.Stderr = os.Stderr
+	if err := resize.Run(); err != nil {
+		return errors.Errorf("resizing image: %q", err)
+	}
+
+	return nil
+}
+
+func (v *MachineVM) setRootful(rootful bool) error {
```
Comment (baude) [inline pkg/machine/qemu/machine.go:1437]: good for a unittest?

**Comment by baude** [comment]
the machine tests just merged.  can you add some e2e tests in pkg/machine/e2e? I am willing to help/pair program with you on some!

**Comment by ashley-cui** [comment]
Sure, let me take a look at those!

**Comment by ashley-cui** [comment]
Added tests!

File: docs/source/markdown/podman-machine-set.1.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (TomSweeneyRedHat) [reply docs/source/markdown/podman-machine-set.1.md:32]: Hmm, the main thing I was getting at for the qemu link was if someone didn't know what "qemu" was, where would they go?  The man page I was looking at was on arch, missed the fact that we don't have one on our distro.

**Comment by TomSweeneyRedHat** [comment]
LGTM
